### PR TITLE
agrep: fix build failure

### PR DESCRIPTION
**Pull request by momeemt**
changed to compile option '-ansi' and patched Makefile

## Description of changes

agrep failed to build on the [all platforms](https://hydra.nixos.org/build/258221963). It needs to be built with C89 and Unix target, so I added a patch file to the Makefile and changed the compiler cc to gcc ansi.
I also added myself to its maintainers, as it is no longer maintained by anyone.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by RaghavSood** [comment]
Result of `nixpkgs-review pr 310847` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

File: pkgs/tools/text/agrep/default.nix [hunk 1]
```diff
@@ -1,33 +1,41 @@
...
+    sha256 = "sha256-2J4bw5BVZgTEcIn9IuD5Q8/L+8tldDbToDefuxDf85g=";
   };
 
-  # Related: https://github.com/Wikinaut/agrep/pull/11
-  prePatch = lib.optionalString (stdenv.hostPlatform.isMusl || stdenv.isDarwin) ''
-    sed -i '1i#include <sys/stat.h>' checkfil.c newmgrep.c recursiv.c
-  '';
+  nativeBuildInputs = [ gcc ];
+
+  patches = [ ./Makefile.patch ];
```
Comment (AndersonTorres) [inline pkgs/tools/text/agrep/default.nix:24]: A short `# comment` explaining why this patch is needed.

File: pkgs/tools/text/agrep/default.nix [hunk 2]
```diff
@@ -1,33 +1,41 @@
...
-    rev = "eef20411d605d9d17ead07a0ade75046f2728e21";
-    sha256 = "14addnwspdf2mxpqyrw8b84bb2257y43g5ccy4ipgrr91fmxq2sk";
+    sha256 = "sha256-2J4bw5BVZgTEcIn9IuD5Q8/L+8tldDbToDefuxDf85g=";
   };
 
-  # Related: https://github.com/Wikinaut/agrep/pull/11
-  prePatch = lib.optionalString (stdenv.hostPlatform.isMusl || stdenv.isDarwin) ''
-    sed -i '1i#include <sys/stat.h>' checkfil.c newmgrep.c recursiv.c
-  '';
+  nativeBuildInputs = [ gcc ];
```
Comment (AndersonTorres) [inline pkgs/tools/text/agrep/default.nix:22]: Is this needed? Usually GCC is given by `stdenv`.

File: pkgs/tools/text/agrep/default.nix [hunk 3]
```diff
@@ -1,33 +1,41 @@
...
   version = "3.41.5";
 
   src = fetchFromGitHub {
+    inherit rev;
     owner = "Wikinaut";
     repo = "agrep";
-    # This repository has numbered versions, but not Git tags.
-    rev = "eef20411d605d9d17ead07a0ade75046f2728e21";
-    sha256 = "14addnwspdf2mxpqyrw8b84bb2257y43g5ccy4ipgrr91fmxq2sk";
+    sha256 = "sha256-2J4bw5BVZgTEcIn9IuD5Q8/L+8tldDbToDefuxDf85g=";
```
Comment (AndersonTorres) [inline pkgs/tools/text/agrep/default.nix:19]:
```suggestion
    hash = "sha256-2J4bw5BVZgTEcIn9IuD5Q8/L+8tldDbToDefuxDf85g=";
```

File: pkgs/tools/text/agrep/Makefile.patch [hunk 4]
```diff
@@ -0,0 +1,11 @@
+--- Makefile	2024-05-11 22:16:29
++++ Makefile.fixed	2024-05-11 22:16:19
+@@ -14,7 +14,7 @@
+ # use
+ # make CROSS=i686-pc-mingw32-
+ 
+-CC=$(CROSS)gcc
++CC=$(CROSS)gcc -ansi -O3 -w
```
Comment (AndersonTorres) [inline pkgs/tools/text/agrep/Makefile.patch:8]: Maybe this patch can be dispensed and you can set `makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" ];` in default.nix.

P.S.: Does this software respect CFLAGS?

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by RaghavSood** [comment]
> Result of `nixpkgs-review pr 310847` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

nixpkgs-review seems to think this PR removes the agrep package for some reason:

```
1 package removed:
agrep (†3.41.5)
```

**Comment by Sigmanificient** [comment]
- ZHF: https://github.com/NixOS/nixpkgs/issues/309482

Fix the following builds:
- https://hydra.nixos.org/build/258370626
- https://hydra.nixos.org/build/258221963

File: pkgs/tools/text/agrep/Makefile.patch [hunk 5]
```diff
@@ -0,0 +1,11 @@
+--- Makefile	2024-05-11 22:16:29
++++ Makefile.fixed	2024-05-11 22:16:19
```
Comment (Sigmanificient) [inline pkgs/tools/text/agrep/Makefile.patch:2]:
```suggestion
--- a/Makefile	2024-05-11 22:16:29
+++ b/Makefile	2024-05-11 22:16:19
```
then you can remove the `patchFlags = [ "-p0" ];` line

**Review by Sigmanificient — CHANGES_REQUESTED** [review]

File: pkgs/tools/text/agrep/default.nix (hunk 2 above)
Comment (momeemt) [reply pkgs/tools/text/agrep/default.nix:22]: > @nix { "action": "setPhase", "phase": "unpackPhase" }
Running phase: unpackPhase
unpacking source archive /nix/store/3cyk02zdfhniq4i5d7cslhi2zvc077d4-source
source root is source
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: patchPhase
applying patch /nix/store/35zyydwfmm8d5xc6ib329218cm3kyn44-Makefile.patch
patching file Makefile
Hunk #1 succeeded at 14 with fuzz 2.
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "configurePhase" }
Running phase: configurePhase
no configure script, doing nothing
@nix { "action": "setPhase", "phase": "buildPhase" }
Running phase: buildPhase
build flags: SHELL=/nix/store/93haq4cas597hmsmvq3g59rjf02c5frw-bash-5.2p26/bin/bash
gcc -ansi -O3 -w -DMEASURE_TIMES=0 -DAGREP_POINTER=1 -DDOTCOMPRESSED=0 -c -DHAVE_DIRENT_H=1 -DHAVE_SYS_DIR_H=0 -DHAVE_SYS_NDIR_H=0 -DHAVE_NDIR_H=0 -DUTIME=1 -DISO_CHAR_SET=1 -DS_IFLNK=-1 -Dlstat=stat -O3   -c -o follow.o follow.c
/nix/store/93haq4cas597hmsmvq3g59rjf02c5frw-bash-5.2p26/bin/bash: line 1: gcc: command not found
make: *** [<builtin>: follow.o] Error 127

I tried to remove `gcc` from nativeBuildInputs, then it threw not found error...
Should I add something in `stdenv` to use gcc in the build?

**Comment by Sigmanificient** [comment]
Result of `nixpkgs-review pr 310847` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

File: pkgs/tools/text/agrep/default.nix [hunk 6]
```diff
@@ -1,33 +1,41 @@
...
-  makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" ];
-
-  meta = with lib; {
+  meta = {
     description = "Approximate grep for fast fuzzy string searching";
     mainProgram = "agrep";
     homepage = "https://www.tgries.de/agrep/";
-    license = licenses.isc;
-    platforms = with platforms; linux ++ darwin;
+    maintainer = with lib.maintainers; [ momeemt ];
```
Comment (Sigmanificient) [inline pkgs/tools/text/agrep/default.nix:36]:
```suggestion
    maintainers = with lib.maintainers; [ momeemt ];
```

**Review by Sigmanificient — CHANGES_REQUESTED** [review]

File: pkgs/tools/text/agrep/default.nix (hunk 2 above)
Comment (annaleeleaves) [reply pkgs/tools/text/agrep/default.nix:22]: this error is due to removing
```
makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" ];
```

arguments past to make on the command line over-ride the definitions in the Makefile. not all stdenv use gcc as the default compiler.

FWIW, the patch below fixes the build for me on master.

<details><summary>agrep.diff</summary>

```diff
diff --git a/pkgs/tools/text/agrep/default.nix b/pkgs/tools/text/agrep/default.nix
index 39de33badd77..4b79fb2f2546 100644
--- a/pkgs/tools/text/agrep/default.nix
+++ b/pkgs/tools/text/agrep/default.nix
@@ -13,7 +13,7 @@ stdenv.mkDerivation {
   };
 
   # Related: https://github.com/Wikinaut/agrep/pull/11
-  prePatch = lib.optionalString (stdenv.hostPlatform.isMusl || stdenv.isDarwin) ''
+  prePatch = ''
     sed -i '1i#include <sys/stat.h>' checkfil.c newmgrep.c recursiv.c
   '';
   installPhase = ''
@@ -21,7 +21,7 @@ stdenv.mkDerivation {
     install -Dm 444 docs/* -t "$out/doc"
   '';
 
-  makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" ];
+  makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" "CC+=-std=c89" ];
 
   meta = with lib; {
     description = "Approximate grep for fast fuzzy string searching";
```
</details>

File: pkgs/tools/text/agrep/default.nix [hunk 7]
```diff
@@ -1,33 +1,41 @@
...
+
+  patches = [ ./Makefile.patch ];
+  patchFlags = [ "-p0" ];
+
   installPhase = ''
     install -Dm 555 agrep -t "$out/bin"
     install -Dm 444 docs/* -t "$out/doc"
   '';
 
-  makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" ];
```
Comment (AndersonTorres) [inline pkgs/tools/text/agrep/default.nix:24]: Why was this removed?

File: pkgs/tools/text/agrep/default.nix (hunk 6 above)
Comment (momeemt) [reply pkgs/tools/text/agrep/default.nix]: I missed it, thank you.

File: pkgs/tools/text/agrep/default.nix [hunk 8]
```diff
@@ -1,33 +1,42 @@
...
+  meta = {
     description = "Approximate grep for fast fuzzy string searching";
     mainProgram = "agrep";
     homepage = "https://www.tgries.de/agrep/";
-    license = licenses.isc;
-    platforms = with platforms; linux ++ darwin;
+    maintainers = with lib.maintainers; [ momeemt ];
+    changelog = "https://github.com/Wikinaut/agrep/blob/${rev}/CHANGES";
+    license = lib.licenses.isc;
+    platforms = with lib.platforms; linux ++ darwin;
```
Comment (AndersonTorres) [inline pkgs/tools/text/agrep/default.nix:40]:
```suggestion
    platforms = lib.platforms.unix;
```

It looks like they are not restricted to Linux and Darwin.

**Review by AndersonTorres — APPROVED** [review]

**Review by Sigmanificient — APPROVED** [review]

File: pkgs/tools/text/agrep/default.nix [hunk 9]
```diff
@@ -1,33 +1,42 @@
...
+  stdenv,
+  fetchFromGitHub,
+}:
+let
+  # This repository has numbered versions, but not Git tags.
+  rev = "b7d180fe73636740f694ec60c1ffab52b06e7150";
+in
 stdenv.mkDerivation {
   pname = "agrep";
   version = "3.41.5";
```
Comment (natsukium) [inline pkgs/tools/text/agrep/default.nix:12]:
```suggestion
  version = "3.41.5-unstable-2022-03-23";
```
Please add `-unstable-YYYY-MM-DD` when we use HEAD.
https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-naming

File: pkgs/tools/text/agrep/default.nix [hunk 10]
```diff
@@ -1,33 +1,42 @@
...
-    # This repository has numbered versions, but not Git tags.
-    rev = "eef20411d605d9d17ead07a0ade75046f2728e21";
-    sha256 = "14addnwspdf2mxpqyrw8b84bb2257y43g5ccy4ipgrr91fmxq2sk";
+    hash = "sha256-2J4bw5BVZgTEcIn9IuD5Q8/L+8tldDbToDefuxDf85g=";
   };
 
   # Related: https://github.com/Wikinaut/agrep/pull/11
   prePatch = lib.optionalString (stdenv.hostPlatform.isMusl || stdenv.isDarwin) ''
     sed -i '1i#include <sys/stat.h>' checkfil.c newmgrep.c recursiv.c
   '';
```
Comment (natsukium) [inline pkgs/tools/text/agrep/default.nix:24]:
```suggestion
```
This PR has already been merged.
https://github.com/Wikinaut/agrep/commit/b56355bfebd0e3e63379b50c19c03579707bb619

File: pkgs/tools/text/agrep/default.nix [hunk 11]
```diff
@@ -1,33 +1,42 @@
...
+
+  makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" "CC+=-std=c89" ];
+
```
Comment (natsukium) [inline pkgs/tools/text/agrep/default.nix:27]:
```suggestion
  makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" ];

  env.NIX_CFLAGS_COMPILE = "-std=c89";
```
This package looks for CFLAGS, so it can be passed with `NIX_CFLAGS_COMPILE`.

**Review by natsukium — CHANGES_REQUESTED** [review]

File: pkgs/tools/text/agrep/default.nix (hunk 11 above)
Comment (natsukium) [reply pkgs/tools/text/agrep/default.nix:27]: Oops, this seems to break Linux (GCC) builds. 
Please use the following code instead.

```nix
  env = lib.optionalAttrs stdenv.cc.isClang {
    NIX_CFLAGS_COMPILE = "-std=c89";
  };
```

**Review by natsukium — APPROVED** [review]
